### ISSUE TITLE
lbann2 results differ for the same model prototext depending on presence of 2nd model prototext

### ISSUE THREAD

**Issue by DylanMcKinney** [LLNL/lbann#537]
This behavior is a bit hard to explain, and I am unsure if it is a bug. The same model, when run using lbann2 AND with a 2nd model specified, has different results than when only that model is run using lbann or lbann2. To duplicate what I am trying to describe run the following:
`srun -n2 build/gnu.Release.pascal.llnl.gov/install/bin/lbann2 --model=model_zoo/models/lenet_mnist/model_lenet_mnist.prototext --reader=model_zoo/data_readers/data_reader_mnist.prototext --optimizer=model_zoo/optimizers/opt_sgd.prototext --num_epochs=2 --disable_cuda` 

followed by: 
`srun -n2 build/gnu.Release.pascal.llnl.gov/install/bin/lbann2 --model={model_zoo/models/lenet_mnist/model_lenet_mnist.prototext,model_zoo/models/lenet_mnist/model_lenet_mnist.prototext} --reader=model_zoo/data_readers/data_reader_mnist.prototext --optimizer=model_zoo/optimizers/opt_sgd.prototext --num_epochs=2 --disable_cuda`

The behavior I am referring to can be seen in the first model prototext results in the 2 model execution. The first epoch of this execution matches the 1 model prototext execution, however the 2nd epoch (and on) differ. 

Like I said I am unsure if this is a bug, but it causes the lbann2 reload unit test to fail, first seen on 7/11. A glance over the [bamboo commit history](https://lc.llnl.gov/bamboo/browse/LBANN-NIGHTD-X86CPU-358/commit) for that day shows #505 merging, so maybe some of the changes in there caused this (based on nothing other than the fact that it has lbann2 changes).
**Edit:** I am trying to validate this on a clean build but most of my LC resources are down/not cooperating. I will close this if this issue doesn't appear in my eventual clean build test

**Comment by DylanMcKinney** [comment]
We [reinitialize](https://github.com/LLNL/lbann/blob/develop/model_zoo/lbann2.cpp#L163 ) random state before beginning the first model prototext execution when a second model proto is present. This isn't necessarily a 'bug', it is just a reproducibility thing. So I don't know that it needs fixing.

**Comment by ndryden** [comment]
Yeah, this was probably caused by #505, which messed with some of the details of initialization. I thought I replicated the functionality as it was before, but it's certainly possible I missed something.

Part of this also stems from us having global random number generator state, which can interact badly when we run multiple models and care about reproducibility. I'm unsure exactly what semantics we want in terms of RNG state.

**Comment by JaeseungYeom** [comment]
I have a similar issue but not sure if these are different or same. I was able to use lbann2 to train Siamese triplet model with a pair of model prototext files (which are identical) on surface thanks to #505.
However, essentially the same runs on pascal using the develop branch after #505 is merged did not learn well. I wish I can verify this again on surface, but currently I cannot run on surface.
